File: backtracking/word_search.py
def dfs(r, c, i, word, board) -> bool:
    if i == len(word):
        return True
    if (
        r >= len(board) or r < 0 or
        c >= len(board[0]) or c < 0 or
        board[r][c] != word[i]
        or board[r][c]  == '*'
    ):
        return False
        
    board[r][c] = "*"
    result = (
        dfs(r + 1, c, i + 1, word, board) or 
        dfs(r - 1, c, i + 1, word, board) or 
        dfs(r, c + 1, i + 1, word, board) or 
        dfs(r, c - 1, i + 1, word, board)
    )
    board[r][c] = word[i]
    return result    

# Grid m * n, word -> bool
# Returns if word exists in grid
def search(board: list[list[int]], word: str) -> bool:
    n = len(board)
    m = len(board[0])
    
    for i in range(n):
        for j in range(m):
            if dfs(i, j, 0, word, board):
                return True
    return False


def test():
    print(search([["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], "ABCCED"))  # True
    print(search([["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], "SEE"))  # True
    print(search([["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], "ABCB"))  # False    
    print(search([["C","A","A"],["A","A","A"],["B","C","D"]], "AAB"))  # False


test()

File: backtracking/word_search.go
package main

import "fmt"

func exist(board [][]byte, word string) bool {
	m := len(board)
	n := len(board[0])

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if dfs(board, word, i, j, 0) == true {
				return true
			}
		}
	}
	return false
}

func dfs(board [][]byte, word string, i, j, k int) bool {
	if k == len(word) {
		return true
	}
	if i < 0 || j < 0 || i == len(board) || j == len(board[0]) || word[k] != board[i][j] {
		return false
	}
	board[i][j] = '#'
	res := dfs(board, word, i+1, j, k+1) || dfs(board, word, i-1, j, k+1) || dfs(board, word, i, j+1, k+1) || dfs(board, word, i, j-1, k+1)
	board[i][j] = word[k]
	return res
}

func main() {
	fmt.Println("Fuck you")
}
